File: 01_Language/01_Functions/go/src/functions/doubleval.go
package functions

import (
	"strconv"
	"strings"
)

func Doubleval(s string) (float64, error) {
	if len(s) == 0 {
		return 0, nil
	}
	var res string
	var (
		digit bool
		sign  bool
		dot   bool
	)
	for _, c := range s {
		if c >= '0' && c <= '9' {
			res += string(c)
			digit = true
		} else if c == '+' || c == '-' {
			if digit || sign || dot {
				break
			}
			res += string(c)
			sign = true
		} else if c == '.' {
			if dot {
				break
			}
			if !digit {
				res += "0"
				digit = true
			}
			res += string(c)
			dot = true
		} else if c == ' ' {
			if digit || sign || dot {
				break
			}
		} else {
			if digit || sign || dot {
				break
			}
			return 0, nil
		}
	}
	if !digit {
		return 0, nil
	}
	if strings.HasSuffix(res, ".") {
		res = res[:len(res)-1]
	}
	return strconv.ParseFloat(res, 64)
}

File: 01_Language/01_Functions/python/doubleval.py
# coding: utf-8

def doubleval(s):
    if s is None or len(s) == 0:
        return 0.0
    res = ""
    digit = False
    sign = False
    dot = False
    for c in s:
        if '0' <= c <= '9':
            res += c
            digit = True
        elif c == '+' or c == '-':
            if digit or sign or dot:
                break
            res += c
            sign = True
        elif c == '.':
            if dot:
                break
            if not digit:
                res += '0'
                digit = True
            res += c
            dot = True
        elif c == ' ':
            if digit or sign or dot:
                break
        else:
            if digit or sign or dot:
                break
            return 0.0
    if not digit:
        return 0.0
    if res.endswith("."):
        res = res[:-1]
    return float(res)


if __name__ == '__main__':
    print(doubleval("122.34343The"))
    print(doubleval("+122.34343."))
    print(doubleval("-122.34343.1"))
    print(doubleval("-.34343.1"))
    print(doubleval("The122.34343"))
